pkg/gardenlet/controller/extensions/extensions_sync_control.go

```diff
@@ -0,0 +1,138 @@
+	workerHasSynced          cache.InformerSynced
+
+	infrastructureExtensionQueue workqueue.RateLimitingInterface
+	workerExtensionQueue         workqueue.RateLimitingInterface
+
+	waitGroup              sync.WaitGroup
+	numberOfRunningWorkers int
+	workerCh               chan int
+	informerShutdown       chan struct{}
+}
+
+// NewExtensionsSyncController creates new controller that syncs extensions states to ShootSnapshot
+func NewExtensionsSyncController(parentCtx context.Context, gardenClient, seedClient kubernetes.Interface, extensionsSeedClient gardenextensionsclient.Interface, extensionsInformers gardenextensionsinformers.SharedInformerFactory, config *config.GardenletConfiguration, log *logrus.Entry, recorder record.EventRecorder) *ExtensionStateSyncController {
+	var (
+		infraInformer        = extensionsInformers.Extensions().V1alpha1().Infrastructures()
+		workerInformer       = extensionsInformers.Extensions().V1alpha1().Workers()
+		infrastructuresList  = infraInformer.Lister()
+		workersList          = workerInformer.Lister()
+		childCtx, cancelFunc = context.WithCancel(parentCtx)
+	)
+
+	extensionController := &ExtensionStateSyncController{
+		k8sGardenClient:              gardenClient,
+		seedClient:                   seedClient,
+		extensionsSeedClient:         extensionsSeedClient,
+		extensionsInformers:          extensionsInformers,
+		infrastructuresList:          infrastructuresList,
+		log:                          log,
+		workerExtensionQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "worker-extension-controllers"),
+		infrastructureExtensionQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "infra-extension-controllers"),
+		recorder:                     recorder,
+		workersList:                  workersList,
+		waitGroup:                    sync.WaitGroup{},
+		context:                      childCtx,
+		cancelFunc:                   cancelFunc,
+		workerCh:                     make(chan int),
+	}
+
+	extensionsInformers.Start(extensionController.informerShutdown)
+
+	infraInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    enqueueOnAdd(extensionController.infrastructureExtensionQueue),
+		UpdateFunc: enqueueOnUpdate(extensionController.infrastructureExtensionQueue),
+	})
+
+	workerInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    enqueueOnAdd(extensionController.workerExtensionQueue),
+		UpdateFunc: enqueueOnUpdate(extensionController.workerExtensionQueue),
+	})
+
+	extensionController.infrastructuresHasSynced = infraInformer.Informer().HasSynced
+	extensionController.workerHasSynced = workerInformer.Informer().HasSynced
+
+	return extensionController
+}
+
+// Run starts the sync controller
+func (e *ExtensionStateSyncController) Run(extensionsSyncWorkersCount int) error {
+	if !cache.WaitForCacheSync(e.context.Done(), e.infrastructuresHasSynced, e.workerHasSynced) {
```

Comment:
Let's create a new WithTimeout context here and pass that to the WaitForCacheSync func.
```
timeoutCtx, cancel := context.WithTimeout(e.context, time.Minute*2)
defer cancel()
```